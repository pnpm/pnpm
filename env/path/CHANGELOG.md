# @pnpm/env.path

## 1.0.0

### Major Changes

- 26b065c: Initial release.
