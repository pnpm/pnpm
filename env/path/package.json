{
  "name": "@pnpm/env.path",
  "version": "1.0.0",
  "description": "Functions for changing PATH env variable",
  "funding": "https://opencollective.com/pnpm",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "exports": {
    ".": "./lib/index.js"
  },
  "engines": {
    "node": ">=18.12"
  },
  "scripts": {
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/env/path",
  "keywords": [
    "pnpm9",
    "pnpm",
    "env"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/env/path#readme",
  "dependencies": {
    "path-name": "catalog:"
  },
  "devDependencies": {
    "@pnpm/env.path": "workspace:*"
  }
}
