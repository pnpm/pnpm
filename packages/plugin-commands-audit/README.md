# @pnpm/plugin-commands-audit

> pnpm commands for dependencies audit

[![npm version](https://img.shields.io/npm/v/@pnpm/plugin-commands-audit.svg)](https://www.npmjs.com/package/@pnpm/plugin-commands-audit)

## Installation

```sh
pnpm add @pnpm/plugin-commands-audit
```

## Testing

In order to have stable tests, the responses from the registry are mocked.
To update the mocks, run:

```
pnpm run update-responses
```

## License

MIT
