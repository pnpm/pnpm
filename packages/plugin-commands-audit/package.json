{
  "name": "@pnpm/plugin-commands-audit",
  "version": "5.0.1",
  "description": "pnpm commands for dependencies audit",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "engines": {
    "node": ">=12.17"
  },
  "scripts": {
    "lint": "eslint -c ../../eslint.json src/**/*.ts test/**/*.ts",
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "rimraf lib tsconfig.tsbuildinfo && tsc --build && pnpm run lint -- --fix"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/master/packages/plugin-commands-audit",
  "keywords": [
    "pnpm",
    "audit"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/master/packages/plugin-commands-audit#readme",
  "devDependencies": {
    "@pnpm/test-fixtures": "workspace:*",
    "@pnpm/types": "workspace:7.3.0",
    "@types/ramda": "0.27.39",
    "@types/zkochan__table": "npm:@types/table@6.0.0",
    "load-json-file": "^6.2.0",
    "nock": "12.0.3",
    "strip-ansi": "^6.0.0",
    "tempy": "^1.0.0"
  },
  "dependencies": {
    "@pnpm/audit": "workspace:2.1.6",
    "@pnpm/cli-utils": "workspace:0.6.10",
    "@pnpm/config": "workspace:12.4.1",
    "@pnpm/constants": "workspace:5.0.0",
    "@pnpm/error": "workspace:2.0.0",
    "@pnpm/lockfile-file": "workspace:4.1.0",
    "@pnpm/read-project-manifest": "workspace:2.0.4",
    "@zkochan/table": "^1.0.0",
    "chalk": "^4.1.0",
    "ramda": "^0.27.1",
    "render-help": "^1.0.1"
  },
  "funding": "https://opencollective.com/pnpm"
}
