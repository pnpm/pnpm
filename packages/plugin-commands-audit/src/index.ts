import * as audit from './audit'

export { audit }
