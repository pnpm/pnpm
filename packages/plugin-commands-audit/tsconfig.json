{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../typings/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../privatePackages/test-fixtures"
    },
    {
      "path": "../audit"
    },
    {
      "path": "../cli-utils"
    },
    {
      "path": "../config"
    },
    {
      "path": "../constants"
    },
    {
      "path": "../error"
    },
    {
      "path": "../lockfile-file"
    },
    {
      "path": "../read-project-manifest"
    },
    {
      "path": "../types"
    }
  ]
}
