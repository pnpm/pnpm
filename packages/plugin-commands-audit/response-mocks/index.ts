import path from 'path'
import loadJsonFile from 'load-json-file'

// eslint-disable-next-line
const response1 = loadJsonFile.sync<any>(path.join(__dirname, 'response1.json'))
// eslint-disable-next-line
const response2 = loadJsonFile.sync<any>(path.join(__dirname, 'response2.json'))
// eslint-disable-next-line
const response3 = loadJsonFile.sync<any>(path.join(__dirname, 'response3.json'))

export { response1, response2, response3 }
