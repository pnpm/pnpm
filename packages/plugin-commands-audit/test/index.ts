import path from 'path'
import { audit } from '@pnpm/plugin-commands-audit'
import nock from 'nock'
import stripAnsi from 'strip-ansi'

const skipOnNode10 = process.version.split('.')[0] === 'v10' ? test.skip : test

// The audits give different results on Node 10, for some reason
skipOnNode10('audit', async () => {
  const { output, exitCode } = await audit.handler({
    dir: path.join(__dirname, 'fixtures/has-vulnerabilities'),
    registries: {
      default: 'https://registry.npmjs.org/',
    },
  })
  expect(exitCode).toBe(1)
  expect(stripAnsi(output)).toMatchSnapshot()
})

test('audit --dev', async () => {
  const { output, exitCode } = await audit.handler({
    dir: path.join(__dirname, 'fixtures/has-vulnerabilities'),
    dev: true,
    production: false,
    registries: {
      default: 'https://registry.npmjs.org/',
    },
  })

  expect(exitCode).toBe(1)
  expect(stripAnsi(output)).toMatchSnapshot()
})

test('audit --audit-level', async () => {
  const { output, exitCode } = await audit.handler({
    auditLevel: 'moderate',
    dir: path.join(__dirname, 'fixtures/has-vulnerabilities'),
    registries: {
      default: 'https://registry.npmjs.org/',
    },
  })

  expect(exitCode).toBe(1)
  expect(stripAnsi(output)).toMatchSnapshot()
})

test('audit: no vulnerabilities', async () => {
  const { output, exitCode } = await audit.handler({
    dir: path.join(__dirname, '../../../fixtures/has-outdated-deps'),
    registries: {
      default: 'https://registry.npmjs.org/',
    },
  })

  expect(stripAnsi(output)).toBe('No known vulnerabilities found\n')
  expect(exitCode).toBe(0)
})

test('audit --json', async () => {
  const { output, exitCode } = await audit.handler({
    dir: path.join(__dirname, 'fixtures/has-vulnerabilities'),
    json: true,
    registries: {
      default: 'https://registry.npmjs.org/',
    },
  })

  const json = JSON.parse(output)
  expect(json.metadata).toBeTruthy()
  expect(exitCode).toBe(1)
})

test.skip('audit does not exit with code 1 if the found vulnerabilities are having lower severity then what we asked for', async () => {
  const { output, exitCode } = await audit.handler({
    auditLevel: 'high',
    dir: path.join(__dirname, 'fixtures/has-vulnerabilities'),
    dev: true,
    registries: {
      default: 'https://registry.npmjs.org/',
    },
  })

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toBe(`1 vulnerabilities found
Severity: 1 moderate`)
})

test('audit does not exit with code 1 if the registry responds with a non-200 reponse and ignoreRegistryErrors is used', async () => {
  const registry = 'https://registry-error.com'
  nock(registry)
    .post('/-/npm/v1/security/audits')
    .reply(500, { message: 'Something bad happened' })
  const { output, exitCode } = await audit.handler({
    dir: path.join(__dirname, 'fixtures/has-vulnerabilities'),
    dev: true,
    ignoreRegistryErrors: true,
    production: false,
    registries: {
      default: registry,
    },
  })

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toBe('The audit endpoint (at https://registry-error.com/-/npm/v1/security/audits) responded with 500: {"message":"Something bad happened"}')
})
