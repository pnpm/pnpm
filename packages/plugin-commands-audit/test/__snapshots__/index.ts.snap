// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`audit --audit-level 1`] = `
"┌─────────────────────┬───────────────────────────────────┐
│ critical            │ Command Injection                 │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ nodemailer                        │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <6.4.16                           │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=6.4.16                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1708 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────────────────────┐
│ critical            │ Improper Verification of Cryptographic Signature │
├─────────────────────┼──────────────────────────────────────────────────┤
│ Package             │ xmlhttprequest-ssl                               │
├─────────────────────┼──────────────────────────────────────────────────┤
│ Vulnerable versions │ <1.6.1                                           │
├─────────────────────┼──────────────────────────────────────────────────┤
│ Patched versions    │ >=1.6.1                                          │
├─────────────────────┼──────────────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1746                │
└─────────────────────┴──────────────────────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Insufficient Entropy              │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ cryptiles                         │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <4.1.2                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=4.1.2                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1464 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Denial of Service                 │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ http-proxy                        │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <1.18.1                           │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.18.1                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1486 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────────────────────────────────┐
│ high                │ Remote Memory Exposure                                        │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ Package             │ bl                                                            │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ Vulnerable versions │ <1.2.3 || >2.0.0 < 2.2.1 || >=3.0.0 <3.0.1 || >= 4.0.0 <4.0.3 │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ Patched versions    │ >=1.2.3 <2.0.0 || >=2.2.1 <3.0.0 || >=3.0.1 <4.0.0 || >=4.0.3 │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1555                             │
└─────────────────────┴───────────────────────────────────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Server-Side Request Forgery       │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ axios                             │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <0.21.1                           │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=0.21.1                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1594 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬────────────────────────────────────────────────────┐
│ high                │ netmask npm package vulnerable to octal input data │
├─────────────────────┼────────────────────────────────────────────────────┤
│ Package             │ netmask                                            │
├─────────────────────┼────────────────────────────────────────────────────┤
│ Vulnerable versions │ <2.0.1                                             │
├─────────────────────┼────────────────────────────────────────────────────┤
│ Patched versions    │ >=2.0.1                                            │
├─────────────────────┼────────────────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1658                  │
└─────────────────────┴────────────────────────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Arbitrary Code Injection          │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ xmlhttprequest-ssl                │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <1.6.2                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.6.2                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1665 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Command Injection                 │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ lodash                            │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <4.17.21                          │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=4.17.21                         │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1673 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Arbitrary Code Execution          │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ underscore                        │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ >=1.3.2 <1.12.1                   │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.12.1                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1674 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Path traversal                    │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ url-parse                         │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <1.5.0                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.5.0                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1678 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬─────────────────────────────────────────┐
│ high                │ Resource exhaustion in socket.io-parser │
├─────────────────────┼─────────────────────────────────────────┤
│ Package             │ socket.io-parser                        │
├─────────────────────┼─────────────────────────────────────────┤
│ Vulnerable versions │ <3.3.2 || =3.4.0                        │
├─────────────────────┼─────────────────────────────────────────┤
│ Patched versions    │ >=3.3.2 <3.4.0 || >=3.4.1               │
├─────────────────────┼─────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1763       │
└─────────────────────┴─────────────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────┐
│ moderate            │ Prototype Pollution              │
├─────────────────────┼──────────────────────────────────┤
│ Package             │ hoek                             │
├─────────────────────┼──────────────────────────────────┤
│ Vulnerable versions │ <= 4.2.0 || >= 5.0.0 < 5.0.3     │
├─────────────────────┼──────────────────────────────────┤
│ Patched versions    │ > 4.2.0 < 5.0.0 || >= 5.0.3      │
├─────────────────────┼──────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/566 │
└─────────────────────┴──────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────┐
│ moderate            │ Memory Exposure                  │
├─────────────────────┼──────────────────────────────────┤
│ Package             │ tunnel-agent                     │
├─────────────────────┼──────────────────────────────────┤
│ Vulnerable versions │ <0.6.0                           │
├─────────────────────┼──────────────────────────────────┤
│ Patched versions    │ >=0.6.0                          │
├─────────────────────┼──────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/598 │
└─────────────────────┴──────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────┐
│ moderate            │ Denial of Service                │
├─────────────────────┼──────────────────────────────────┤
│ Package             │ axios                            │
├─────────────────────┼──────────────────────────────────┤
│ Vulnerable versions │ <0.18.1                          │
├─────────────────────┼──────────────────────────────────┤
│ Patched versions    │ >=0.18.1                         │
├─────────────────────┼──────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/880 │
└─────────────────────┴──────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ moderate            │ Insecure Default Configuration    │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ socket.io                         │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <2.4.0                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=2.4.0                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1609 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────────┐
│ moderate            │ Regular expression denial of service │
├─────────────────────┼──────────────────────────────────────┤
│ Package             │ glob-parent                          │
├─────────────────────┼──────────────────────────────────────┤
│ Vulnerable versions │ <5.1.2                               │
├─────────────────────┼──────────────────────────────────────┤
│ Patched versions    │ >=5.1.2                              │
├─────────────────────┼──────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1751    │
└─────────────────────┴──────────────────────────────────────┘
24 vulnerabilities found
Severity: 7 low | 5 moderate | 10 high | 2 critical"
`;

exports[`audit --dev 1`] = `
"┌─────────────────────┬───────────────────────────────────┐
│ high                │ Server-Side Request Forgery       │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ axios                             │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <0.21.1                           │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=0.21.1                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1594 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────┐
│ moderate            │ Denial of Service                │
├─────────────────────┼──────────────────────────────────┤
│ Package             │ axios                            │
├─────────────────────┼──────────────────────────────────┤
│ Vulnerable versions │ <0.18.1                          │
├─────────────────────┼──────────────────────────────────┤
│ Patched versions    │ >=0.18.1                         │
├─────────────────────┼──────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/880 │
└─────────────────────┴──────────────────────────────────┘
2 vulnerabilities found
Severity: 1 moderate | 1 high"
`;

exports[`audit 1`] = `
"┌─────────────────────┬───────────────────────────────────┐
│ critical            │ Command Injection                 │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ nodemailer                        │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <6.4.16                           │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=6.4.16                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1708 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────────────────────┐
│ critical            │ Improper Verification of Cryptographic Signature │
├─────────────────────┼──────────────────────────────────────────────────┤
│ Package             │ xmlhttprequest-ssl                               │
├─────────────────────┼──────────────────────────────────────────────────┤
│ Vulnerable versions │ <1.6.1                                           │
├─────────────────────┼──────────────────────────────────────────────────┤
│ Patched versions    │ >=1.6.1                                          │
├─────────────────────┼──────────────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1746                │
└─────────────────────┴──────────────────────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Insufficient Entropy              │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ cryptiles                         │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <4.1.2                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=4.1.2                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1464 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Denial of Service                 │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ http-proxy                        │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <1.18.1                           │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.18.1                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1486 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────────────────────────────────┐
│ high                │ Remote Memory Exposure                                        │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ Package             │ bl                                                            │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ Vulnerable versions │ <1.2.3 || >2.0.0 < 2.2.1 || >=3.0.0 <3.0.1 || >= 4.0.0 <4.0.3 │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ Patched versions    │ >=1.2.3 <2.0.0 || >=2.2.1 <3.0.0 || >=3.0.1 <4.0.0 || >=4.0.3 │
├─────────────────────┼───────────────────────────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1555                             │
└─────────────────────┴───────────────────────────────────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Server-Side Request Forgery       │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ axios                             │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <0.21.1                           │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=0.21.1                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1594 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬────────────────────────────────────────────────────┐
│ high                │ netmask npm package vulnerable to octal input data │
├─────────────────────┼────────────────────────────────────────────────────┤
│ Package             │ netmask                                            │
├─────────────────────┼────────────────────────────────────────────────────┤
│ Vulnerable versions │ <2.0.1                                             │
├─────────────────────┼────────────────────────────────────────────────────┤
│ Patched versions    │ >=2.0.1                                            │
├─────────────────────┼────────────────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1658                  │
└─────────────────────┴────────────────────────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Arbitrary Code Injection          │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ xmlhttprequest-ssl                │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <1.6.2                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.6.2                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1665 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Command Injection                 │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ lodash                            │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <4.17.21                          │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=4.17.21                         │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1673 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Arbitrary Code Execution          │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ underscore                        │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ >=1.3.2 <1.12.1                   │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.12.1                          │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1674 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ high                │ Path traversal                    │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ url-parse                         │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <1.5.0                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=1.5.0                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1678 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬─────────────────────────────────────────┐
│ high                │ Resource exhaustion in socket.io-parser │
├─────────────────────┼─────────────────────────────────────────┤
│ Package             │ socket.io-parser                        │
├─────────────────────┼─────────────────────────────────────────┤
│ Vulnerable versions │ <3.3.2 || =3.4.0                        │
├─────────────────────┼─────────────────────────────────────────┤
│ Patched versions    │ >=3.3.2 <3.4.0 || >=3.4.1               │
├─────────────────────┼─────────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1763       │
└─────────────────────┴─────────────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────┐
│ moderate            │ Prototype Pollution              │
├─────────────────────┼──────────────────────────────────┤
│ Package             │ hoek                             │
├─────────────────────┼──────────────────────────────────┤
│ Vulnerable versions │ <= 4.2.0 || >= 5.0.0 < 5.0.3     │
├─────────────────────┼──────────────────────────────────┤
│ Patched versions    │ > 4.2.0 < 5.0.0 || >= 5.0.3      │
├─────────────────────┼──────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/566 │
└─────────────────────┴──────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────┐
│ moderate            │ Memory Exposure                  │
├─────────────────────┼──────────────────────────────────┤
│ Package             │ tunnel-agent                     │
├─────────────────────┼──────────────────────────────────┤
│ Vulnerable versions │ <0.6.0                           │
├─────────────────────┼──────────────────────────────────┤
│ Patched versions    │ >=0.6.0                          │
├─────────────────────┼──────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/598 │
└─────────────────────┴──────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────┐
│ moderate            │ Denial of Service                │
├─────────────────────┼──────────────────────────────────┤
│ Package             │ axios                            │
├─────────────────────┼──────────────────────────────────┤
│ Vulnerable versions │ <0.18.1                          │
├─────────────────────┼──────────────────────────────────┤
│ Patched versions    │ >=0.18.1                         │
├─────────────────────┼──────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/880 │
└─────────────────────┴──────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ moderate            │ Insecure Default Configuration    │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ socket.io                         │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <2.4.0                            │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=2.4.0                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1609 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────────┐
│ moderate            │ Regular expression denial of service │
├─────────────────────┼──────────────────────────────────────┤
│ Package             │ glob-parent                          │
├─────────────────────┼──────────────────────────────────────┤
│ Vulnerable versions │ <5.1.2                               │
├─────────────────────┼──────────────────────────────────────┤
│ Patched versions    │ >=5.1.2                              │
├─────────────────────┼──────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1751    │
└─────────────────────┴──────────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────────┐
│ low                 │ Regular Expression Denial of Service │
├─────────────────────┼──────────────────────────────────────┤
│ Package             │ timespan                             │
├─────────────────────┼──────────────────────────────────────┤
│ Vulnerable versions │ >=0.0.0                              │
├─────────────────────┼──────────────────────────────────────┤
│ Patched versions    │ <0.0.0                               │
├─────────────────────┼──────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/533     │
└─────────────────────┴──────────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────────┐
│ low                 │ Regular Expression Denial of Service │
├─────────────────────┼──────────────────────────────────────┤
│ Package             │ braces                               │
├─────────────────────┼──────────────────────────────────────┤
│ Vulnerable versions │ <2.3.1                               │
├─────────────────────┼──────────────────────────────────────┤
│ Patched versions    │ >=2.3.1                              │
├─────────────────────┼──────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/786     │
└─────────────────────┴──────────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ low                 │ Prototype Pollution               │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ minimist                          │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <0.2.1 || >=1.0.0 <1.2.3          │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=0.2.1 <1.0.0 || >=1.2.3         │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1179 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ low                 │ Validation Bypass                 │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ kind-of                           │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ >=6.0.0 <6.0.3                    │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=6.0.3                           │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1490 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬───────────────────────────────────┐
│ low                 │ Prototype Pollution               │
├─────────────────────┼───────────────────────────────────┤
│ Package             │ lodash                            │
├─────────────────────┼───────────────────────────────────┤
│ Vulnerable versions │ <4.17.19                          │
├─────────────────────┼───────────────────────────────────┤
│ Patched versions    │ >=4.17.19                         │
├─────────────────────┼───────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1523 │
└─────────────────────┴───────────────────────────────────┘
┌─────────────────────┬──────────────────────────────────────┐
│ low                 │ Regular Expression Denial of Service │
├─────────────────────┼──────────────────────────────────────┤
│ Package             │ redis                                │
├─────────────────────┼──────────────────────────────────────┤
│ Vulnerable versions │ >=2.6.0 <3.1.1                       │
├─────────────────────┼──────────────────────────────────────┤
│ Patched versions    │ >=3.1.1                              │
├─────────────────────┼──────────────────────────────────────┤
│ More info           │ https://npmjs.com/advisories/1662    │
└─────────────────────┴──────────────────────────────────────┘
24 vulnerabilities found
Severity: 7 low | 5 moderate | 10 high | 2 critical"
`;
